### PR TITLE
[DebugInfo][RemoveDIs] Add a DPValue implementation for instcombine sinking

### PR DESCRIPTION
**Pull request by jmorse**
Hi (and pre-emptive apologies for an essay),

This patch re-implements debug-info salvaging for instcombine sinking using DPValues, the replacement for dbg.value intrinsics. Unfortunately it does it in an ugly way, while I'll elaborate below. The objective here is to match what instcombine currently does, i.e. when we sink an instruction from one block to an immediate successor, the latest assignments to variables using that instruction value sink with it. I've refactor'd the dbg.value implementation into it's own function and added a DPValue one, as it's too much stuff to really observe side-by-side efficiently.

I dislike this sinking-salvaging behaviour, because it's trying to maintain an up-to-date fully-accurate varible-to-values mapping in the middle of compilation -- which is something that I don't think is feasible. As demonstrated with instruction referencing in the codegen backend, we'd be better off tolerating some use-before-defs by dbg.values and just specifying their meaning. We could technically do that right now and not sink dbg.values at all: however most architectures don't support instruction referencing yet, so they'd lose variable location coverage. Another reason I don't like this is because it's not fully accurate: debug intrinsics will re-order with intrinsics for the same variable, but which use a different LLVM-IR Value. (And fixing that would introduce a large compile-time cost I believe).

When replacing dbg.values with DPValues, we need to preserve the existing behaviour. However there's a major problem -- with dbg.value intrinsics there's a total order of instructions in a block that we can rely on for ordering dbg.values and finding the last assignment to a variable. This exists for DPValues, but it's harder to use, because the order of DPValues is hidden in a secondary list attached to each instruction, and we can't easily find the order of two DPValues that are attached to the same instruction. I've resorted to producing an instruction/Variable count-densemap to identify situations where there are duplicate sinking DPValues attached to the same instruction, then finding the last DPValue within that instruction, and recording that we should only sink that one. This is expensive, but should only fire rarely. compile time tracker:

  http://llvm-compile-time-tracker.com/compare.php?from=90525125421300d9d1b6bf55288bd1871855d35d&to=eabd2a9d6922dd82b59497b769bc0a160e69c811&stat=instructions:u

And there's even more difficulty! Because we don't get and use the total order of DPValues in the block, when we clone and sink some DPValues, they won't happen in the same order as in dbg.value mode. That changes the order of creation (equivalent to the use-list order in a ValueAsMetadata), which causes other bits of LLVM to process dbg.values/DPValues in different orders. None of this causes a functional difference, but because DW_TAG_variables etc in DWARF are ordered by the order in which we encounter debug-info records in the instruction stream, we lose the ability to produce an identical binary. This is something we just have to swallow IMO: one of the primary reasons for the RemoveDIs project is not needlessly putting debug-info into blocks and making it part of the total-order-of-everything, but output consistency with dbg.value mode relies on that, so we have to give up on identical-outputs at some point.

### CODE REVIEW

**Comment by jmorse** [comment]
(Immediate observation; the compile-time difference observe on the O3 configuration is likely because we want an early-out when there's no debug-info, which I'll add).

File: llvm/test/DebugInfo/instcombine-sink-latest-assignment.ll [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+declare ptr @_ZNK4llvm6MCInst10getOperandEj(i32) local_unnamed_addr
+
+declare i64 @_ZNK4llvm25SmallVectorTemplateCommonINS_9MCOperandEvE4sizeEv()
+
+!llvm.dbg.cu = !{!0}
+!llvm.module.flags = !{!3}
+
+!0 = distinct !DICompileUnit(language: DW_LANG_C_plus_plus_14, file: !1, isOptimized: true, runtimeVersion: 0, emissionKind: FullDebug, enums: !2, retainedTypes: !2, globals: !2, imports: !2, splitDebugInlining: false, nameTableKind: None)
+!1 = !DIFile(filename: "/fast/fs/llvm34/lib/MC/MCObjectStreamer.cpp", directory: "/fast/fs/build34llvmstage", checksumkind: CSK_MD5, checksum: "43f3adff5ece50116e446307bd92824d")
```
Comment (SLTozer) [inline llvm/test/DebugInfo/instcombine-sink-latest-assignment.ll:57]: Nit, we can remove the useless metadata e.g. checksums.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp [hunk 2]
```diff
@@ -4266,19 +4285,140 @@ bool InstCombinerImpl::tryToSinkInstruction(Instruction *I,
...
+      DPValuesToSalvage.push_back(DPV);
+
+  // Fetch a second collection, of DPValues in the source block that we're going
+  // to sink.
+  SmallVector<DPValue *> DPValuesToSink;
+  for (DPValue *DPV : DPValuesToSalvage)
+    if (DPV->getParent() == SrcBlock)
+      DPValuesToSink.push_back(DPV);
+
+  // Sort DPValues according to their position in the block. This is a partial
```
Comment (SLTozer) [inline llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:4317]: Have you tested the performance of biting the O(n) bullet and inspecting the ordering of DPValues attached to a single instruction directly as part of the `Order` lambda? Worst case it'll be worse than what's here, but I'm curious how the overhead of using several map/set structures (even `Small` ones) compare in the typical case.

**Review by SLTozer — COMMENTED** [review]
Code seems good and clear, a couple questions and suggested change inline.

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp (hunk 2 above)
Comment (jmorse) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:4317]: Numbers: http://llvm-compile-time-tracker.com/compare.php?from=90525125421300d9d1b6bf55288bd1871855d35d&to=5c1b0a6c7d28637f3e0ac059e5af561a96396365&stat=instructions%3Au

No regression in stage1-ReleaseLTO-g, there are some differences in the non-debuginfo modes, I'm not sure why that is really. Something else to keep our eyes on I guess.

**Comment by jmorse** [comment]
Pushed revisions in def72c5, and merged master into this branch which appears to have made github throw it's cookies.

**Comment by jmorse** [comment]
https://github.com/jmorse/llvm-project/commit/def72c586d84551a80b95623258496aa5f7947a2

File: llvm/lib/Transforms/InstCombine/InstructionCombining.cpp (hunk 2 above)
Comment (SLTozer) [reply llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:4413]: I actually thinking about the performance of removing all the filtermap/duplicate detection logic and just iterating over getDbgValueRange() to determine whether A or B comes first when they're attached to the same instruction - I submitted a quick test to the compile time tracker[0] though and it looks like it's neutral for most cases and a slowdown in the bad case (testing specifically with RemoveDIs enabled), so I'm happy with this approach.
[0] http://llvm-compile-time-tracker.com/compare.php?from=319280746b199b35c49798383b6dd4c01286c5ff&to=c29a7ff328a4148aabd7e147f2bfc04e9801fcb6&stat=instructions:u

**Review by SLTozer — APPROVED** [review]
LGTM with latest changes.

**Pull request merged**